signers/signer-solana/src/utils/simulate.ts

```diff
@@ -24,16 +32,55 @@ export async function simulateTransaction(
        * If you are getting an invalid account error, make sure that you have the input mint account to actually swap from.
        */
       console.error('Simulation Error:', { err, logs });
-      const message =
-        (logs?.length || 0) > 0
-          ? logs?.[logs?.length - 1]
-          : JSON.stringify(err);
 
-      throw new SignerError(
+      throw getSimulationError(err as SimulationResponseError, logs);
+    }
+  }
+}
+
+function getSimulationError(
```

Comment:
I just refactored your code. (didn't test it)

```ts
function getSimulationError(
  error: SimulationResponseError,
  logs: string[] | null
) {
  let message =
    (logs?.length || 0) > 0 ? logs?.[logs?.length - 1] : JSON.stringify(error);

  // trying to detect common errors (e.g. insufficient fund or slippage error)
  // We could probably remove this code after upgrading solana/web3 lib to v2
  if (
    typeof error === 'object' &&
    error?.InstructionError?.[0] === INSTRUCTION_INDEX &&
    typeof error?.InstructionError?.[1] === 'object'
  ) {
    if (
      error?.InstructionError?.[1]?.Custom === INSUFFICIENT_FUNDS_ERROR_CODE
    ) {
      const insufficentLamportErrorMessage = logs?.find((log) =>
        log?.toLowerCase()?.includes('insufficient lamports')
      );
      message = insufficentLamportErrorMessage
        ? insufficentLamportErrorMessage
        : 'Insufficient funds';
    } else if (error?.InstructionError?.[1]?.Custom === SLIPPAGE_ERROR_CODE) {
      message = 'Slippage error';
    }
  }

  return new SignerError(
    SignerErrorCode.SEND_TX_ERROR,
    undefined,
    `Simulation failed: ${message}`
  );
}

```

Reply:
Applied